### ISSUE TITLE
Choosing a test layout

### ISSUE THREAD

**Issue by hackebrot** [hackebrot/qutebrowser#7]
https://pytest.org/latest/goodpractises.html#choosing-a-test-layout-import-rules

**Comment by The-Compiler** [comment]
I've seen this and actually was about to ask about this :)

For unittest I had to add the `__init__.py` files to get the tests run - I guess this means currently (with pytest and tox) the tests run against the version in the cwd, not against the installed package, right?

I'd be happy with moving the tests directory out of the qutebrowser package, there's no need to ship them with the package anyways - and there are some html files for manual testing there already, which makes this rather confusing:

```
test
└── html
    ├── jsconfirm.html
    └── jsprompt.html
qutebrowser/test
└── browser
    └── ...
```

I guess something like this would make more sense, right?

```
test
├── html
│   ├── jsconfirm.html
│   └── jsprompt.html
└── unittests
    ├── browser
    └── ...
qutebrowser
└── ...
```

**Comment by nicoddemus** [comment]
@The-Compiler, I completely agree. :smile:

I would suggest though that we rename this files at the latest stage in our adoption process, to avoid possible merge conflicts down the road (although my experience is that Git usually handles those just fine, but you never know). Thoughts?

**Comment by The-Compiler** [comment]
I'd actually suggest to do this as early as possible (i.e. before or after merging #9) - then I can cherry-pick/pull that change to `The-Compiler/master` immediately. This means if I make changes to the tests as well before merging with your repo, merge conflicts there would be avoided (or a lot easier to handle). :smile:

**Comment by hackebrot** [comment]
I think we should make this decision at the earliest possible as it is actually more than just moving files. If @The-Compiler decides to go with the separated test layout we need to make sure that we can run the suite on the installed package.

That being said we also need to move the code from the `__init__.py` files to according `conftest.py` files.

Do you think we can sort this out any time soon :question:

**Comment by nicoddemus** [comment]
If @The-Compiler plans to cherry-pick this change directly, then I agree we should this earliest rather than latest. :smile:

The only code I've seen in `__init__.py` files is in `qutebrowser.test.__init__`... that code creates a `QApplication` object but that is already handled by `pytest-qt`, so this can be safely removed.

So all that is left is moving the files from `qutebrowser` package and into a `test` folder in the same level, without `__init__` files. Perhaps that can be done by @The-Compiler instead? Because @hackebrot and I don't have access to your C.I. server to ensure things are working correctly.

**Comment by The-Compiler** [comment]
I think there are two changes which need to be done to get rid of the `__init__.py`:
- Make `pytest-qt` actually initialize a `QApplication` - does this happen automatically already with it installed, or is there something which should be added to the test code?
- Ideally find some solution for `log.py`, see #8 - maybe for now that can be just moved to a `conftest.py`?

I don't think you have to worry about the CI too much. If you open a PR here, I can easily test it on the CI and give you feedback - then we merge it here, and then I can cherry-pick it. But really, I'm pretty sure if it runs locally it'll run on the CI as well. :pray:

**Comment by nicoddemus** [comment]
> Make pytest-qt actually initialize a QApplication - does this happen automatically already with it installed

It will create a `QApplication` the first time it sees the `qtbot` fixture, which is problem right now because not all of your tests use the `qtbot` fixture yet, but see below.

> Ideally find some solution for log.py, see #8 - maybe for now that can be just moved to a conftest.py?

Yes, we can move the `init` and related code into a session-scoped fixture... like this:

``` python
# contents of test/conftest.py
import pytest
@pytest.fixture(scope='session', autouse=True)
def app_and_logging(qapp):
    from .log import init
    init()
```

This will ensure to initialize this fixture only once per session and for all tests. Just tested it, and all seems good.

**Comment by nicoddemus** [comment]
Argh, I had already tested and made the changes, so why not [push it](https://github.com/hackebrot/qutebrowser/commit/751b62e344d85baf8182565425d79be230a08d78)? :smile:

About the `qapp` fixture: that's supplied by `pytest-qt`, but it will only be created when it executes the first test which uses the `qtbot` fixture. Making a autouse session-fixture which uses it will guarantee that it will stay alive for all tests.

**Comment by nicoddemus** [comment]
Btw, tomorrow I will move the tests as we discussed, if that's alright with you @hackebrot? :smile:

**Comment by nicoddemus** [comment]
@hackebrot, does @The-Compiler have write access to our fork? I was thinking @The-Compiler  himself would merge our PRs into our `master` when he finds it acceptable and has no more questions about the features being presented. How does that sound to both of you?

**Comment by hackebrot** [comment]
Sure @nicoddemus go ahead :+1:

**Comment by hackebrot** [comment]
I just added @The-Compiler to the collaborators :white_check_mark:

**Comment by nicoddemus** [comment]
Guys, I just noticed that once we move the tests outside of the package, the modules `qutebrowser.test.helpers` and `qutebrowser.test.stubs` can't be imported by the tests anymore; we would have to convert the functionality from those modules into fixtures and change the tests that use them (about 10+ files). Should I go ahead?

**Comment by The-Compiler** [comment]
I think that'd be a good start for me to see how fixtures work and are used - so yeah, if you feel like, please go ahead!

Then as soon as this has landed in your master I think I'll do a first merge into my master.

**Comment by nicoddemus** [comment]
OK, working on this then.

**Comment by hackebrot** [comment]
I think this is totally up to @The-Compiler.

Please note that I am currently submitting several PRs which are using the original layout. We may merge them first. #10 

I suggest to create Pull Requests from this forks master to `qutebrowser` instead of cherry-picking as we should have at least a description in the history of the project that briefly explains the concepts introduced.

**Comment by nicoddemus** [comment]
Agree about merging the other PRs first; I will deal with any merge conflicts (if any).

**Comment by hackebrot** [comment]
The new tests reside in a new directory. I do not expect any conflicts.

**Comment by hackebrot** [comment]
Closing as we merged #16. :smiley:
